File: 130.java
import java.util.LinkedList;
import java.util.Queue;

public class Solution {
    private void bfs(char[][] board, int row, int col) {
        int rows = board.length, cols = board[0].length;
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{row, col});
        board[row][col] = 'N';

        while (!queue.isEmpty()) {
            int[] current = queue.remove();
            int r = current[0], c = current[1];
            int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

            for (int[] d : directions) {
                int newRow = r + d[0], newCol = c + d[1];
                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && board[newRow][newCol] == 'O') {
                    queue.add(new int[]{newRow, newCol});
                    board[newRow][newCol] = 'N';
                }
            }
        }
    }

    public void solve(char[][] board) {
        if (board == null || board.length == 0 || board[0].length == 0) {
            return;
        }
        int rows = board.length;
        int cols = board[0].length;

        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if ((row == 0 || row == rows - 1 || col == 0 || col == cols - 1) && board[row][col] == 'O') {
                    bfs(board, row, col);
                }
            }
        }

        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if (board[row][col] == 'O') {
                    board[row][col] = 'X';
                } else if (board[row][col] == 'N') {
                    board[row][col] = 'O';
                }
            }
        }
    }
}

File: 130.py
from typing import List

class Solution:
    def solve(self, board: List[List[str]]) -> None:
        if not board:
            return
        rows, cols = len(board), len(board[0])

        def bfs(row, col):
            queue = [(row, col)]
            # mark the borders "N" (not capturable)
            board[row][col] = "N"
            while queue:
                row, col = queue.pop(0)
                indices = [(row - 1, col), (row + 1, col), (row, col - 1), (row, col + 1)]
                for row, col in indices:
                    if row in range(rows) and col in range(cols) and board[row][col] == "O":
                        queue.append((row, col))
                        board[row][col] = "N"

        # Start BFS from all border "O"s
        for row in range(rows):
            for col in range(cols):
                if (row == 0 or row == rows - 1 or col == 0 or col == cols - 1) and board[row][col] == "O":
                    bfs(row, col)

        # print(board)
        # Flip the capturable "O"s to "X"s and revert non-capturable "O"s back
        for row in range(rows):
            for col in range(cols):
                if board[row][col] == "O":
                    board[row][col] = "X"
                elif board[row][col] == "N":
                    board[row][col] = "O"

board = [["X","X","X","X"],["X","O","O","X"],["X","X","O","X"],["X","O","X","X"]]
board2 = [["O","O","O"],["O","O","O"],["O","O","O"]]
s = Solution()
s.solve(board)
print(board)
